Add I/F to get window/workDoneProgress

micchy326 opened the pull request:
- sample screen shot
![simplescreenrecorder-2020-12-28_00 27 30](https://user-images.githubusercontent.com/23257067/103174258-eee84280-48a3-11eb-9448-9da9ee8d372e.gif)

Implement lsp#get_progress() for statusline plugins.
I initially created the following PR, but decided to introduce it to CORE.
https://github.com/mattn/vim-lsp-settings/pull/361

relating PR
prabirshrestha/vim-lsp#970

vim-airline sample implementation
https://github.com/micchy326/vim-airline

- remaining tasks
  * [X]  Write documents
  * [X] Consider the case where multiple servers are registered.
         https://github.com/mattn/vim-lsp-settings/pull/361#discussion_r548898953

prabirshrestha commented on autoload/lsp/internal/work_done_progress.vim:
```diff
@@ -0,0 +1,45 @@
+          \ lsp#callbag#filter({x->has_key(x, 'response') && has_key(x['response'], 'method')
+          \  && x['response']['method'] ==# '$/progress' && has_key(x['response'], 'params')
+          \  && has_key(x['response']['params'], 'value') && type(x['response']['params']['value']) == type({})}),
+          \  lsp#callbag#subscribe({'next': function('s:handle_work_done_progress')})
```
Can we change to this so it aligns with https://github.com/prabirshrestha/vim-lsp/pull/978/files

```vim
 \  lsp#callbag#subscribe({'next': {x->s:handle_work_done_progress(x['server'], x['response']}})
```

This allows the code to be self documented.

micchy326 replied:
Thank you for comment.
I fixed it.

prabirshrestha commented on autoload/lsp/internal/work_done_progress.vim:
```diff
@@ -0,0 +1,45 @@
+endfunction
+
+function! s:handle_work_done_progress(progress) abort
+    let l:value = a:progress['response']['params']['value']
```
Always check for response error.

prabirshrestha commented on autoload/lsp/internal/work_done_progress.vim:
```diff
@@ -0,0 +1,45 @@
+let s:lsp_progress = {
```
Please add a link to  https://microsoft.github.io/language-server-protocol/specifications/specification-current/#progress at the top of the file as a comment.

micchy326 replied:
I fixed it.

prabirshrestha commented on autoload/lsp.vim:
```diff
@@ -1122,6 +1124,10 @@ function! lsp#get_buffer_first_error_line() abort
     return lsp#ui#vim#diagnostics#get_buffer_first_error_line()
 endfunction
 
+function! lsp#get_progress() abort
```
missing documentation.

micchy326 replied:
I wrote documentation.

prabirshrestha commented on autoload/lsp/internal/work_done_progress.vim:
```diff
@@ -0,0 +1,45 @@
+function! lsp#internal#work_done_progress#_enable() abort
+    if !g:lsp_work_done_progress_enabled | return | endif
+
+    let s:Dispose = lsp#callbag#pipe(
```
also add `s:enabled` check like https://github.com/prabirshrestha/vim-lsp/pull/978/files so someone doesn't accidentally subscribe multiple times. Primarily used for unit testing.

micchy326 replied:
Thank you for advice.
I fixed it.

prabirshrestha commented on autoload/lsp/internal/work_done_progress.vim:
```diff
@@ -0,0 +1,45 @@
+          \ )
+endfunction
+
+function! s:handle_work_done_progress(progress) abort
```
do we need `lsp_diagnostics_updated` autocmd?

let us not add it yet if we don't require for vim airline or lightline.

micchy326 replied:
Not necessary at this time.
But we may want it in the future.

Not implemented for now.

micchy326 replied:
Sorry.
Added `lsp_progress_updated` autocmd, which I said was unnecessary before.
So, I attempted several implementations that did not use autocmd.
But, the implementation made users wait since progress notification is very frequently.  

So I added it separately from lsp_diagnostics_updated to reduce the update frequency by the status line plugin side.

prabirshrestha replied:
i mean it should had been `lsp_progress_updated`. what you have is better. thanks.

prabirshrestha reviewed (changes requested):
Added comments.

It would be great if you could also add unit tests as I'm looking into improving the stability of vim-lsp a lot.

prabirshrestha commented on autoload/lsp/internal/work_done_progress.vim:
```diff
@@ -0,0 +1,45 @@
+        let s:lsp_progress['title'] = l:value['title']
+    elseif l:value['kind'] ==# 'report'
+        let s:lsp_progress['messages'] = get(l:value, 'message', '')
+        let s:lsp_progress['percentage'] = get(l:value, 'percentage', '')
```
spec says progress should always be a number or can be missing.

micchy326 replied:
I fixed it.
If 'message' does not exist, use -1.0

micchy326 commented:
Added the feature to output the results of multiple servers.
The sample airline implementation will always switch between new results.
https://github.com/micchy326/vim-airline

-sample screen(clangd+rust-analyzer)
![simplescreenrecorder-2020-12-29_02 23 18](https://user-images.githubusercontent.com/23257067/103232229-43132580-497d-11eb-843b-2e0f16e5a696.gif)

prabirshrestha commented on autoload/lsp/internal/work_done_progress.vim:
```diff
@@ -0,0 +1,62 @@
+        let s:progress_ui = filter(s:progress_ui, {_, x->x['token'] !=# l:token})->insert(l:new)
+    elseif l:value['kind'] ==# 'report'
+        let l:new['messages'] = get(l:value, 'message', '')
+        let l:new['percentage'] = get(l:value, 'percentage', -1.0)
```
spec says percentage should be uinteger.

```vim
if has_key(l:value, 'percentage') | let l:new['percentage'] = l:value['percentage'] | endif

micchy326 replied:
I'm sorry.
I misunderstood the specification, because clangd and rust-analyzer return float.
But they should be 'uinteger' in the future.
microsoft/language-server-protocol#1037

So I fixed to uinteger.

prabirshrestha commented on autoload/lsp/internal/work_done_progress.vim:
```diff
@@ -0,0 +1,62 @@
+      \ 'token': l:token,
+      \ 'title': '',
+      \ 'messages': '',
+      \ 'percentage': -1,
```
spec says it needs to be `uinteger` so let us ignore this and not add if percentage is not there.

micchy326 replied:
I fixed it.

prabirshrestha commented on autoload/lsp/internal/work_done_progress.vim:
```diff
@@ -0,0 +1,62 @@
+          \ )
+endfunction
+
+function! s:handle_work_done_progress(server, progress) abort
```
can you rename `progress` as `response`. 

This is one thing i have realized after working in vim script codebase and vim-lsp quite a lot is that in the future it becomes quite difficult to know what the `progress` is since there is no typings. trying to use generic term helps here.

micchy326 replied:
Thank you for advice.
I renamed it.

prabirshrestha reviewed (changes requested):
added few more comments.

prabirshrestha commented on autoload/lsp/internal/work_done_progress.vim:
```diff
@@ -0,0 +1,68 @@
+    if s:enabled | return | endif
+    let s:enabled = 1
+
+    let s:Dispose = lsp#callbag#pipe(
```
we should reset the state of `s:progress_ui` when we enable and disable.

We need to do this when we enable so we get clear state for the first time. and disable so that when `lsp#get_progress()` is called it gets empty array.

prabirshrestha replied:
this is also useful for tests so that you are guaranteed your state is clean before your tests start.

micchy326 replied:
That is true.
I fixed it.

prabirshrestha commented on autoload/lsp/internal/work_done_progress.vim:
```diff
@@ -0,0 +1,68 @@
+endfunction
+
+function! lsp#internal#work_done_progress#_disable() abort
+    if !s:enabled | return | endif
```
reset state `let s:progress_ui  = []`

micchy326 replied:
I fixed it.

prabirshrestha commented on autoload/lsp/internal/work_done_progress.vim:
```diff
@@ -0,0 +1,68 @@
+function! s:handle_work_done_progress(server, response) abort
+    let l:value = a:response['params']['value']
+    " Add the server name to distinguish the server
+    let l:token = a:server . ':' . a:response['params']['token']
```
can we avoid this and instead in filter check both token and server?

micchy326 replied:
No.
When the same token is received from different servers, both should be handled correctly.
Therefore, we cannot filter it.

prabirshrestha replied:
I mean why not do something like this so you the token is the same as the original one.

```vim
let s:progress_ui = filter(s:progress_ui, {_, x->x['token'] !=# l:token && x['server'] !=# a:server })
```

Let me know if I'm missing something.

micchy326 replied:
sorry. I was confused with callbag#filter.
I fixed it.

prabirshrestha commented on test/lsp/utils/work_done_progress.vimspec:
```diff
@@ -0,0 +1,101 @@
+    End
+
+    It should be able to subscribe to $progress stream
+        call lsp#internal#work_done_progress#_disable()
```
why not move this to `Before each` and `After each` so you don't need to do it every where.

micchy326 replied:
I fixed it.

prabirshrestha commented on test/lsp/utils/work_done_progress.vimspec:
```diff
@@ -0,0 +1,101 @@
+        call lsp#internal#work_done_progress#_enable()
+    End
+
+    After all
```
`all` not required.

micchy326 replied:
I fixed it.

prabirshrestha reviewed (changes requested):
looking great! added few more small comments and should be good to merge.

prabirshrestha commented on doc/vim-lsp.txt:
```diff
@@ -1644,7 +1667,7 @@ You can check if semantic highlighting is enabled by running: >
 <
 Before you can use the semantic highlighting feature, you will also need to
-link language server semantic scopes to Vim highlight groups. For example, you
+link language server semantic scopes to Vim highlight groups. For example, yo
```
you

micchy326 replied:
oh sorry...

micchy326 commented:
I fixed comments.
And added `lsp_progress_updated` 

Please see this comment.
https://github.com/prabirshrestha/vim-lsp/pull/979#discussion_r550435146

prabirshrestha commented:
@micchy326 commented. what you have is the way to go.

prabirshrestha commented on autoload/lsp/internal/work_done_progress.vim:
```diff
@@ -0,0 +1,71 @@
+      \ 'server': a:server,
+      \ 'token': l:token,
+      \ 'title': '',
+      \ 'messages': '',
```
spec has message without s.

https://microsoft.github.io/language-server-protocol/specification#workDoneProgressBegin

micchy326 replied:
oh sorry.
i fixed it.

prabirshrestha commented:
This is awesome. This is going to make it lot better for me to use methods like rust analyzer.

Currently here is what I'm doing when using lightline.

```vim
let g:lightline = {
  \ 'colorscheme': 'solarized',
  \ 'active': {
  \   'left': [
  \     ['mode', 'paste'],
  \     ['lsp_progress'],
  \   ],
  \ },
  \ 'component_function': {
  \   'lsp_progress': 'MyLspProgress'
  \ },
  \ }

let g:lsp_work_done_progress_enabled = 1
function! MyLspProgress() abort
  let l:progress = lsp#get_progress()
  if empty(l:progress) | return '' | endif
  let l:progress = l:progress[len(l:progress) - 1]
  return l:progress['server'] . ': ' . l:progress['message']
endfunction

augroup my_lightline_lsp
  autocmd!
  autocmd User lsp_progress_updated call lightline#update()
augroup END
```

Wondering if we should have `lsp#get_progress_last_text()`

```vim
  \ 'component_function': {
  \   'lsp_progress': 'lsp#get_progress_last_text'
  \ },
```

Not sure about other statusline but it would be the vimrc look smaller.

polyzen commented:
Works great in Crystalline as well: https://github.com/polyzen/dotfiles/commit/0895e62efa196ea281b99757e6b6b7741a53e391

micchy326 commented:
I wrote plugin for lightline.
https://github.com/micchy326/lightline-lsp-progress
